### PR TITLE
GpuAdvancedSubtensor

### LINKED ISSUE #3166: Convert AdvancedSubtensor to GpuAdvancedSubtensor1 on the GPU
@lamblin you told me you know how to make this. Can you describe it here so that someone else have a change to make it?

### PR DESCRIPTION
**Pull request by yingzha**
Fixes #3166, #4097 
@abergeron, sorry for opening a pull request, the previous one has some conflicts and I failed to rebase it. 
I made some changes, it still has the problem as I said in the previous ticket #4097, 
it got a TypeError saying Variables do not support boolean operations and I didn't figure it out.

Could you help me check it?

### CODE REVIEW

File: theano/sandbox/gpuarray/tests/test_subtensor.py [hunk 1]
```diff
@@ -56,3 +56,45 @@ def test_advinc_subtensor1():
         rep = xval.copy()
         rep[[0, 2]] += yval
         assert numpy.allclose(rval, rep)
+
+
+class G_subtensor(test_subtensor.TestAdvancedSubtensor):
+     def shortDescription(self):
+        return None
+     def __init__(self, name):
+         test_subtensor.T_subtensor.__init__(
```
Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_subtensor.py]: Should be TestAdvancedSubtensor

File: theano/sandbox/gpuarray/subtensor.py [hunk 2]
```diff
@@ -484,6 +484,122 @@ def c_code_cache_version(self):
...
+
+            idx_1 = []
+            idx_2 = []
+            idx_3 = []
+            for i in range(a+1, b):
+                if isinstance(ind[i], slice):
+                    idx_1.append(k)
+                    idx_2.append(ind[i])
+                    k += 1
+                elif ind[i] == None:
```
Comment (abergeron) [inline theano/sandbox/gpuarray/subtensor.py]: Compare with `is`

File: theano/sandbox/gpuarray/subtensor.py [hunk 3]
```diff
@@ -484,6 +484,122 @@ def c_code_cache_version(self):
...
+            dimshuffle_info.extend(idx_3)
+            new_ind += idx_2
+            dimshuffle_info.extend(idx_1)
+
+            for i in range(b, len(ind)):
+                if isinstance(ind[i], slice):
+                    dimshuffle_info.append(k)
+                    new_ind.append(ind[i])
+                    k += 1
+                elif ind[i] == None:
```
Comment (abergeron) [inline theano/sandbox/gpuarray/subtensor.py]: Also here with `is`

File: theano/sandbox/gpuarray/subtensor.py [hunk 4]
```diff
@@ -484,6 +484,122 @@ def c_code_cache_version(self):
...
+                    new_ind.append(slice(None))
+
+            return dimshuffle_info, new_ind, valid_end
+
+        (dimshuffle_idx, new_ind,
+                end_) = get_indices(start, end, idx)
+        x = x.transpose(*dimshuffle_idx)
+        #step 3: partial flattening
+        start_ = start
+        import pdb; pdb.set_trace()
```
Comment (abergeron) [inline theano/sandbox/gpuarray/subtensor.py]: Stray debug statement.

File: theano/sandbox/gpuarray/tests/test_subtensor.py [hunk 5]
```diff
@@ -56,3 +56,43 @@ def test_advinc_subtensor1():
...
+    def shortDescription(self):
+        return None
+    def __init__(self, name):
+        test_subtensor.TestAdvancedSubtensor.__init__(
+            self, name,
+            shared=gpuarray_shared_constructor,
+            sub=GpuAdvancedSubtensor,
+            inc_sub=GpuIncSubtensor,
+            adv_sub1=GpuAdvancedSubtensor1,
+            adv_incsub1=GpuAdvancedIncSubtensor1,
```
Comment (abergeron) [inline theano/sandbox/gpuarray/tests/test_subtensor.py]: This test class doesn't take adv_sub1 and adv_incsub1 as arguments.

File: theano/sandbox/gpuarray/subtensor.py [hunk 6]
```diff
@@ -484,6 +484,121 @@ def c_code_cache_version(self):
...
+
+            return dimshuffle_info, new_ind, valid_end
+
+        (dimshuffle_idx, new_ind,
+                end_) = get_indices(start, end, idx)
+        x = x.transpose(*dimshuffle_idx)
+        # step 3: partial flattening
+        start_ = start
+        shape = (x.shape[: start_] +
+                 (tensor.prod(x.shape[start: end_]),) +
```
Comment (abergeron) [inline theano/sandbox/gpuarray/subtensor.py]: Don't use symbolic operations in perform.

File: theano/sandbox/gpuarray/subtensor.py [hunk 7]
```diff
@@ -484,6 +484,121 @@ def c_code_cache_version(self):
...
+                 x.shape[end_:])
+        input_flat = tensor.reshape(x, shape)
```
Comment (abergeron) [inline theano/sandbox/gpuarray/subtensor.py]: idem

File: theano/sandbox/gpuarray/subtensor.py [hunk 8]
```diff
@@ -484,6 +484,121 @@ def c_code_cache_version(self):
...
+        # step 4: build the strides
+        strides = [1]
+        for i in range(start_, end_-1)[::-1]:
+            stride = x.shape[i+1] * strides[-1]
+            strides.append(stride)
+        # step 5: build the indices into x_flat
+        items = [new_ind[i] if isinstance(new_ind[i], numpy.ndarray)
+                 else 0 for i in range(start_, end_)]
+        new_idx = tensor.sum([i * j for i,j
```
Comment (abergeron) [inline theano/sandbox/gpuarray/subtensor.py]: idem

**Comment by abergeron** [comment]
I've noted where you use symbolic operations in your perform code and this is probably the cause of this error:

```
ERROR: Test the advancedsubtensor on gpu.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/anakha/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/anakha/ext/theano/theano/sandbox/gpuarray/tests/test_subtensor.py", line 96, in test_adv_subtensor
    rval = f(idx1_val, idx2_val)
  File "/home/anakha/ext/theano/theano/compile/function_module.py", line 871, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/home/anakha/ext/theano/theano/gof/link.py", line 148, in raise_with_op
    if getattr(ipt[0], "size", -1) <= 5:
  File "/home/anakha/ext/theano/theano/tensor/var.py", line 74, in __nonzero__
    return self.__bool__()
  File "/home/anakha/ext/theano/theano/tensor/var.py", line 90, in __bool__
    "Variables do not support boolean operations. This "
TypeError: Variables do not support boolean operations. This can happen if you do a logical operation (<, <=, >, <=, ==, !=) between a numpy.ndarray and a Theano tensorvariable. Due to NumPy implementation before NumPy 1.8, we cannot make the Python syntax work when the ndarray is on the left, and this results in this error. To work around that, either call theano.tensor.{lt,le,eq,ne,gt,ge}(ndarray, tensor), or use the Python syntax with the Theano tensor on the left. Or update to NumPy 1.8 or above.
```

File: theano/gpuarray/subtensor.py [hunk 9]
```diff
@@ -447,6 +447,122 @@ def c_code_cache_version(self):
...
+                    new_ind.append(slice(None))
+
+            return dimshuffle_info, new_ind, valid_end
+
+        (dimshuffle_idx, new_ind,
+                end_) = get_indices(start, end, idx)
+        x = x.transpose(*dimshuffle_idx)
+        #step 3: partial flattening
+        start_ = start
+        import pdb; pdb.set_trace()
```
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: This probably shouldn't be here.

File: theano/gpuarray/subtensor.py [hunk 10]
```diff
@@ -447,6 +447,121 @@ def c_code_cache_version(self):
...
+            stride = x.shape[i + 1] * strides[-1]
+            strides.append(stride)
+        # step 5: build the indices into x_flat
+        items = [new_ind[i] if isinstance(new_ind[i], numpy.ndarray)
+                 else 0 for i in range(start_, end_)]
+        new_idx = numpy.sum([i * j for i, j
+                             in zip(items, strides[::-1])],
+                            axis=0)
+        # step 6: advanced slicing
+        out_flat = input_flat.take(new_idx.flatten())
```
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: I just discovered that GpuArrays don't expose `take()`.  So we should probably do that.  I'll have a PR for that tomorrow.

**Comment by abergeron** [comment]
The code looks ok, I'll have to add the `take1()` method to pygpu and then try the tests.

**Comment by abergeron** [comment]
You'll need this PR for libgpuarray: https://github.com/Theano/libgpuarray/pull/188

Also, please get rid of the merge commit.

**Comment by yingzha** [comment]
Thanks!

**Comment by yingzha** [comment]
@abergeron , I'm trying to deal with the "new axis" case where the indice list is something like[id1, id2, None, None, ...], since there is no dimshuffle for gpuarray, I have to use reshape + tranpose to do the same thing as dimshuffle. Is there any easier way to tackle with this case?

**Comment by abergeron** [comment]
No, you have to use reshape + transpose.  That's also what the dimshuffle op is implemented with.

**Comment by yingzha** [comment]
@abergeron, I pushed a new version and it works with new axis now. Please inform me if I did something wrong.

File: theano/gpuarray/subtensor.py [hunk 11]
```diff
@@ -461,6 +461,130 @@ def c_code_cache_version(self):
...
+        input_flat = x.reshape(shape)
+        # step 4: build the strides
+        strides = [1]
+        for i in range(0, end_ - 1)[::-1]:
+            stride = x.shape[i + 1] * strides[-1]
+            strides.append(stride)
+        # step 5: build the indices into x_flat
+        items = [new_ind[i] if isinstance(new_ind[i], numpy.ndarray)
+                 else 0 for i in range(0, end_)]
+        new_idx = numpy.sum([i * j for i, j
```
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: This should use the python sum and not pass any axis argument.

**Comment by abergeron** [comment]
Finished in #4763

**Pull request closed without merging**